RFC-0025-sdpa-optm-cpu.md

```diff
@@ -0,0 +1,77 @@
+| float32  | TRUE     | Inference | Unfused         | 3.463              |         |
+|          |          |           | Flash attention | 1.662              | 2.083634|
+| bfloat16 | FALSE    | Inference | Unfused         | 1.203              |         |
+|          |          |           | Flash attention | 1.154              | 1.042461|
+| bfloat16 | TRUE     | Inference | Unfused         | 1.543              |         |
+|          |          |           | Flash attention | 1.154              | 1.337088|
+| float32  | FALSE    | Training  | Unfused         | 54.938             |         |
+|          |          |           | Flash attention | 23.029             | 2.385601|
+| float32  | TRUE     | Training  | Unfused         | 58.266             |         |
+|          |          |           | Flash attention | 17.835             | 3.266947|
+| bfloat16 | FALSE    | Training  | Unfused         | 18.924             |         |
+|          |          |           | Flash attention | 18.886             | 1.002012|
+| bfloat16 | TRUE     | Training  | Unfused         | 21.08              |         |
+|          |          |           | Flash attention | 14.172             | 1.48744 |
+
+### Stable Diffusion
+Following model's BKM.
+
+| Dtype    | SDPA                    | Throughput (fps) | Speedup SDPA | Total Time (ms) | Speedup |
+| -------- | --------                | -------          | -------      | -------         | ------- |
+| float32  | Unfused                 | 1.63             |              | 1139            |         |
+|          | Flash attention         | 1.983            | 1.216564     | 547.488         | 2.080411|
+| bfloat16 | Flash attention in IPEX | 4.784            |              | 429.051         |         |
+|          | Flash attention         | 4.857            | 1.015259     | 408.823         | 1.049479|
+
+## **Related PRs**
+Flash attention Implementation:
+*   [#103826 Flash attention forward](https://github.com/pytorch/pytorch/pull/103826)
+*   [#104693 Flash attention backward](https://github.com/pytorch/pytorch/pull/103826)
+*   [#104863 enable bfloat16 on flash attention](https://github.com/pytorch/pytorch/pull/104863)
+
+SDPA selecting function:
+*   [#105131 add sdpa choice and UT](https://github.com/pytorch/pytorch/pull/105131)
+
+Some additional works:
+*   [#104583 enable mkl’s bfloat16 gemm on PT](https://github.com/pytorch/pytorch/pull/104583)
+*   [#104584 expand functional utils on CPU vectorization path](https://github.com/pytorch/pytorch/pull/104584)
+
+## **Discussion**
+
+For the SDPA optimization, there are two things that needed to be discussed and I hope to have your precious opinions.
+
+One is about the util functions for SDPA selection. The current util functions are under the CUDA folder, i.e. `transformers/cuda/sdp_utils`. For CPU, we have similar functions in `transformers/sdp_utils_cpp` (see #105131). It is good to know whether we need to make them a unified API.
```

Comment:
There exist compilation issues when the files in CUDA folder include the files in non-CUDA folder, and vice versa. I'll spend more time to figure out the solution.

Reply:
I think that this can be resolved, there is functionality that is needed for cuda dispatching that requires cuda utils to be built. I think though that the common utils can be abstracted out and shared between the two

Reply:
Thanks, would do it.

Reply:
The common utils abstraction has been done in [#105131](https://github.com/pytorch/pytorch/pull/105131).